### PR TITLE
Command occ 'user:report -> count user dirs system wide

### PR DESCRIPTION
**Pull request by AlexAndBear**
## Description
Bugfix: Command occ 'user:report' might not count 'user directories' correctly

Before this PR the underlying function of 'user:report' just looked up in the
'datadirectory' set in 'config.php'.
This implies that user directories which are symlinks or even not in the '
datadirectory', have not been taken into account.
With this PR we check if the user's home path exist and increase the
'user directories' count.

## Related Issue

- May fix https://github.com/owncloud/enterprise/issues/4742

## How Has This Been Tested?

- Create user admin and user2
- Run `occ user:report`
- `| user directories | 2 |` -> **correct**
---
- Run `php occ user:move-home user2 /home/test ` and remove the old home
- Run `occ user:report`
- `| user directories | 1 |` -> **not correct**
---
- Apply this patch
- Run `occ user:report`
- `| user directories | 2 |` -> **correct**

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:

- [x] Code changes
- [ ] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [x] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

### CODE REVIEW

**Comment by JanAckermann** [comment]
@pmaier1 FYI

**Comment by JanAckermann** [comment]
Note: This also fixed the issue that folders that are in the `datadirectory` and are non-user folders will taken into account

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>apiProxySmoke-8-8-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/32497/172/1

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>apiProxySmoke-8-3-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/32497/167/1

**Comment by jvillafanez** [comment]
Some things to test:
* After syncing all LDAP users (assume 500-1000 users) and log in with some of them (2-4)
* Home folder outside data directory (the user_ldap app has that option)
* Object storage as primary storage

The first 2 should be working with the patch I think (needs confirmation), but I'm not sure about the last one.

**Comment by JanAckermann** [comment]
Note: We could also improve the performance by calling countForSeen users instead of using coundForAllUsers, 
what do you think @jvillafanez

File: tests/Core/Command/User/ReportTest.php [hunk 1]
```diff
@@ -67,6 +67,10 @@ protected function setUp(): void {
 		list($storage) = $view->resolvePath('');
 		/** @var $storage Storage */
 
+		/**
+		 * Create some folders in the 'datadirectory'
+		 * which should not been count in as user directories
```
Comment (phil-davis) [inline tests/Core/Command/User/ReportTest.php:72]:
```suggestion
		 * which should not be counted as user directories
```

File: changelog/unreleased/39254 [hunk 2]
```diff
@@ -0,0 +1,12 @@
+Bugfix: Command occ 'user:report' might not count 'user directories' correctly
+
+Before this PR the underlying function of 'user:report' just looked up in the
+'datadirectory' set in 'config.php'.
+This implies that user directories which are symlinks or even not in the '
+datadirectory', have not been taken into account.
```
Comment (phil-davis) [inline changelog/unreleased/39254:6]:
```suggestion
This implies that user directories which are symlinks or even not in the
'datadirectory', have not been taken into account.
```

File: changelog/unreleased/39254 [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+With this PR we check if the user's home path exist and increase the
```
Comment (phil-davis) [inline changelog/unreleased/39254:7]:
```suggestion
With this PR we check if the user's home path exists and increase the
```

**Comment by JanAckermann** [comment]
@jvillafanez 
> Home folder outside data directory (the user_ldap app has that option)
Where do I find this setting?

**Comment by jvillafanez** [comment]
If I remember correctly, in the advanced tab of the wizard there is the "user home folder naming rule". You can set an LDAP attribute there, such as "attr:userhome" (the "attr:" prefix is mandatory). The value for the attribute could contain something like "/opt/user01" for the user01. Each user is expected to have a different home folder.

I don't know if there is any specific attribute you can use for this, so maybe you need to use any unused one.

You could search for "homeFolderNamingRule" in the user_ldap app to check for additional things.

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>webUIComments-chrome-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/32507/124/1

**Comment by JanAckermann** [comment]
@jvillafanez 
I checked case 1, which works fine.
 I tried to import users with the attr:userhome but my import tool breaks, do you dare to check if case 2 is working?

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>apiWebdavMove2-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/32512/92/1

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>webUISharingPublic1-chrome-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/32512/144/1

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>webUISharingIntGroup2-chrome-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/32512/141/1

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>apiSharingNotificationsToShares-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/32512/100/1

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>webUISharingExt1-10.7.0-chrome-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/32512/155/1

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>apiShareOperationsToShares1-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/32512/68/1

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>webUISharingAutocomp1-chrome-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/32512/138/1

**Comment by ownclouders** [comment]
:boom: Acceptance tests pipeline <strong>apiFederationToShares1-latest-mariadb10.2-php7.4</strong> failed. The build has been cancelled.

https://drone.owncloud.com/owncloud/core/32512/108/1

**Comment by jvillafanez** [comment]
> I tried to import users with the attr:userhome but my import tool breaks, do you dare to check if case 2 is working?

Looks good to me.

Tried using "attr:title" as home attribute, with the following values:
* aaliyah_aañote => /opt/aañote
* aaliyah_bähr => /opt/bähr
* aaliyah_abernathy => /opt/aa/abernathy
* aaliyah_adams => /opt/aa/adams

```
www-data@73f0601c68e3: ~/owncloud # ./occ user:report
+--------------------------+---+
| User Report              |   |
+--------------------------+---+
| OCA\User_LDAP\User_Proxy | 4 |
| OC\User\Database         | 1 |
|                          |   |
| guest users              | 0 |
|                          |   |
| total users              | 5 |
|                          |   |
| user directories         | 5 |
+--------------------------+---+

www-data@73f0601c68e3: ~/owncloud # ls data/
admin  avatars  files_external  index.html  owncloud.log

www-data@73f0601c68e3: ~/owncloud # ls /opt
 aa  'aa'$'\303\261''ote'  'b'$'\303\244''hr'

root@73f0601c68e3: /var/www/owncloud # ls /opt/aa
abernathy  adams
```
Ignore the weird folder name inside "/opt", the container has a posix locale by default. Using en_US.UTF-8 as locale fixes that. Nothing to do there.

It seems we're requiring that all users have the configured home attribute set (if active in the wizard), so I haven't been able to sync all the users. Anyway, I don't expect a wrong count

**Comment by JanAckermann** [comment]
Awesome thx, we are checking s3 primary storage

**Comment by JanAckermann** [comment]
@jvillafanez User reoprt directory count, does not work properly with s3 primary storage, tough it also does not work with master.
I suggest documenting this as a known limitation and leave it as it is, so far.

**Comment by jvillafanez** [comment]
Using a view object should work with S3 as primary storage, but it might break with LDAP home folder.
I'm not sure, but if the S3 home uses relative paths, we could use a view object if the path is relative and check the underlying FS if it's an absolute path. Hopefully we don't need something more smart than that.

**Comment by JanAckermann** [comment]
@jvillafanez We try it with s3, creating a new View and calling view->is_dir($userHomePath) is returning false, any idea?

**Comment by JammingBen** [comment]
I don't think we can work with the View (or any OC filesystem methods) here as the implementation circumvents all this stuff. It "just" speaks directly via the S3 API as far as I can tell.

**Comment by JanAckermann** [comment]
Note: As said in comment https://github.com/owncloud/core/pull/39254#issuecomment-922915319, before this PR, user directory count was also 0 (not correct)

**Review by phil-davis — APPROVED** [review]
LGTM - code is simple and logical.
We have some acceptance tests, which still pass - good. They are "simple" because they only have users whose home dir is the the default data. But IMO that is OK.

**Comment by JanAckermann** [comment]
Merged and opened issue #https://github.com/owncloud/core/issues/39278

**Pull request merged**